Src/Particle/AMReX_ParticleContainerI.H

```diff
@@ -701,87 +723,162 @@ ParticleContainer<NStructReal, NStructInt, NArrayReal, NArrayInt, Allocator>
-                        // Set the metadata (presumably velocity) to the mass-weighted average.
-                        //
-                        for (int i = 1; i < NStructReal; i++)
-                        {
-                            pold.rdata(i) = static_cast<ParticleReal>((old_mass*pold.rdata(i) + new_mass*pnew.rdata(i))/total_mass);
-                        }
-                        pold.rdata(0) = static_cast<ParticleReal>(total_mass);
+                        amrex::Gpu::Atomic::AddNoRet(&partData(iv,1+NStructInt+i), p.idata(NStructInt+i));
                     }
                 }
             }
+        });
 
-            //
-            // Add the aggregated particles to the virtuals.
-            //
-            for (const auto& agg_particle : agg_map)
-            {
-                virts.push_back(agg_particle.second);
-            }
-        }
+    //There may be a better way to ensure virts is the right length
+    virts.resize(imf.sum(0));
+
+    int last_offset = 0;
+    for (MFIter mfi(mf); mfi.isValid(); ++mfi)
+    {
+        const auto bx = mfi.tilebox();
+        const auto partData = mf.array(mfi);
+        const auto imf_arr = imf.array(mfi);
+
+        Gpu::DeviceVector<int> offsets(bx.numPts());
+        auto offsets_ptr = offsets.dataPtr();
+        int next_offset = Scan::ExclusiveSum((int) bx.numPts(),(imf_arr.ptr(bx.smallEnd(),0)),(offsets.dataPtr()),Scan::retSum);
+        auto dst = virts.getParticleTileData();
+        ParallelFor(bx, [=] AMREX_GPU_DEVICE(int i, int j, int k)
+        {
+             if(imf_arr(i,j,k,0)!=0)
+             {
+                 SuperParticleType p;
+                 p.cpu() = 0;
+                 p.id() = VirtualParticleID;
+
+                 //Set rdata(0) first so we can normalize the weighted fields
+                 p.rdata(0) = static_cast<ParticleReal>(partData(i,j,k,AMREX_SPACEDIM));
+                 //Set pos with the normalized weighted field
+                 for (int n = 0; n < AMREX_SPACEDIM; ++n)
+                 {
+                     p.pos(n) = static_cast<ParticleReal>(partData(i,j,k,n) / p.rdata(0));
+                 }
+                 //Set rdata(n>0) with the normalized weighted field for NStructReal
+                 for (int n = 1; n < NStructReal; ++n)
+                 {
+                     p.rdata(n) = static_cast<ParticleReal>(partData(i,j,k,AMREX_SPACEDIM+n) / p.rdata(0));
+                 }
+                 //Set rdata(n>0) with the normalized weighted field for NArrayReal
+                 for (int n = 0; n < NArrayReal; ++n)
+                 {
+                     p.rdata(NStructReal+n) = static_cast<ParticleReal>(partData(i,j,k,AMREX_SPACEDIM+NStructReal+n));
+                 }
+                 //Set idata with the "first" particles idata field for NStructInt
+                 for (int n = 0; n < NStructInt; ++n)
+                 {
+                     p.idata(n) = (long unsigned int) imf_arr(i,j,k,1+n);
```

Comment:
Why do you need to do this cast? Is it to avoid a compiler warning when `NStructInt` is `0`?

Reply:
I believe this cast was from a compiler complaining about casting from signed to unsigned